### PR TITLE
add createSchema and getSchema util methods

### PR DESCRIPTION
**Pull request by maha-hajja**
### Description

Add util functions to create and get schema.

PR for the exported function used from conduit: # https://github.com/ConduitIO/conduit/pull/1683

todos:
- use the proto messages and the proto convert functions from commons (not there yet)
- use the schema service from conduit to get and create the schema for builtin connector. (when schema service is implemented in conduit)

Related to https://github.com/ConduitIO/conduit-processor-sdk/issues/45

### Quick checks:

- [ ] There is no other [pull request](https://github.com/conduitio/conduit-processor-sdk/pulls) for the same update/change.
- [ ] I have written unit tests.
- [ ] I have made sure that the PR is of reasonable size and can be easily reviewed.

### CODE REVIEW

File: wasm/errors.go [hunk 1]
```diff
@@ -30,13 +30,23 @@ const (
...
 var (
 	ErrNoMoreCommands         = NewError(ErrorCodeNoMoreCommands, "no more commands")
 	ErrUnknownCommandRequest  = NewError(ErrorCodeUnknownCommandRequest, "unknown command request")
 	ErrUnknownCommandResponse = NewError(ErrorCodeUnknownCommandResponse, "unknown command response")
 	ErrMemoryOutOfRange       = NewError(ErrorCodeMemoryOutOfRange, "memory out of range")
+
+	ErrSchemaUnmarshal = NewError(ErrorCodeSchemaUnmarshal, "failed to unmarshal the proto request")
+	ErrSchemaMarshal   = NewError(ErrorCodeSchemaMarshal, "failed to marshal the proto response into a schema")
+	ErrSchemaGet       = NewError(ErrorCodeSchemaGet, "error while getting the schema")
+	ErrSchemaCreate    = NewError(ErrorCodeSchemaCreate, "error while creating the schema")
```
Comment (lovromazgon) [inline wasm/errors.go:49]: These are pretty generic errors, I would think about changing them into general-purpose errors that can be reused in any host call. For instance, the umarshalling/marshalling error can happen in any host call. Regarding schema errors, we need more specific ones like "schema not found" when getting a schema or "invalid schema" when creating it. Maybe there's more.

I think we also need to provide equivalent error structs in the root of the repo so the processor can use them. The processor code shouldn't directly use this code, but it needs to be able to know if a schema was not found for instance.

We can pair on this if you want to figure out the proper way to structure this.

File: wasm/util.go [hunk 2]
```diff
@@ -50,3 +54,61 @@ func (u Util) Logger(ctx context.Context) *zerolog.Logger {
 	l := u.logger.With().Ctx(ctx).Logger()
 	return &l
 }
+
+func CreateSchema(req *processorv1.CreateSchemaRequest) (*processorv1.CreateSchemaResponse, error) {
```
Comment (lovromazgon) [inline wasm/util.go:58]: I would move these two functions into `schema.go`.

Comment (maha-hajja) [reply wasm/util.go:58]: there's no file called `schema.go`, do you mean add the file under the schema directory? or maybe create a util file under `schema`?

File: schema/schema_wasm.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+func Get(req *processorv1.GetSchemaRequest) (*processorv1.GetSchemaResponse, error) {
+	schema, err := wasm.GetSchema(req)
+	if err != nil {
+		return nil, err
+	}
+	return schema, nil
+}
+
+func Create(req *processorv1.CreateSchemaRequest) (*processorv1.CreateSchemaResponse, error) {
```
[the remark this answers is not in the document]
Comment (maha-hajja) [reply schema/schema_wasm.go:35]: already in progress! was waiting for the proto conversion code, then realized it's not needed here :D

File: schema/schema_wasm.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+// todo: change these method's fields to accept schema subject and version and so, instead of the proto requests.
+// will be done after the proto conversion methods are added to conduit-commons
+
+func Get(req *processorv1.GetSchemaRequest) (*processorv1.GetSchemaResponse, error) {
+	schema, err := wasm.GetSchema(req)
+	if err != nil {
+		return nil, err
+	}
+	return schema, nil
```
[the remark this answers is not in the document]
Comment (maha-hajja) [reply schema/schema_wasm.go:32]: this function should get some fields and create the proto request first, then return the proto response, so this won't be as simple after the changes

File: wasm/util.go (hunk 2 above)
Comment (lovromazgon) [reply wasm/util.go:58]: I mean to create a file `wasm/schema.go` which contains the code related to the schema wasm host functions.

File: wasm/imports.go [hunk 5]
```diff
@@ -41,3 +41,29 @@ func _commandRequest(ptr unsafe.Pointer, size uint32) uint32
 //
 //go:wasmimport conduit command_response
 func _commandResponse(ptr unsafe.Pointer, size uint32) uint32
+
+// Imports `create_schema` from the host, which created a schema
```
Comment (hariso) [inline wasm/imports.go:45]: Typo: created -> creates.

File: wasm/imports.go [hunk 6]
```diff
@@ -41,3 +41,29 @@ func _commandRequest(ptr unsafe.Pointer, size uint32) uint32
...
+// (2) the size of allocated memory.
+//
+// The return value indicates the size of the allocated request in bytes. If the
+// command is larger than the allocated memory, the caller should reallocate the
+// memory and call `create_schema` again.
+//
+//go:wasmimport conduit create_schema
+func _createSchema(ptr unsafe.Pointer, size uint32) uint32
+
+// Imports `get_schema` from the host, which created a schema
```
Comment (hariso) [inline wasm/imports.go:58]: "which gets a schema"

File: wasm/caller.go [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+type HostFunc func(ptr unsafe.Pointer, size uint32) uint32
+
+type HostCaller struct {
+	Func HostFunc
+}
+
+// Call calls the function from the host 2 times max, is the buffer size is not
+// enough the first time its called, it will be resized the second call.
+// returns the buffer, the command size, and the error.
+func (fc *HostCaller) Call(buf []byte, size int) ([]byte, uint32, error) {
```
Comment (lovromazgon) [inline wasm/caller.go:41]: 3 comments here:
- I don't see a reason for this to be a type method, maybe change this into a static method instead?
- I wouldn't necessarily expose it outside of this package.
- Do we need an explicit `size` parameter? Can we instead just use `len(buf)`?

Something like this maybe:
```suggestion
func hostCall(fn hostFunc, buf []byte) ([]byte, int, error) {
```

File: wasm/command.go [hunk 8]
```diff
@@ -28,30 +26,17 @@ const defaultCommandSize = 1024 // 1kB
...
-			continue                       // try again
-		}
-
-		// parse the command
-		if err := proto.Unmarshal(buffer[:cmdSize], cmdReq); err != nil {
-			return fmt.Errorf("failed unmarshalling %v bytes into proto type: %w", cmdSize, err)
-		}
-		return nil
+	commandRequestCaller := HostCaller{Func: _commandRequest}
+	buf, cmdSize, err := commandRequestCaller.Call(buffer, cap(buffer))
```
Comment (lovromazgon) [inline wasm/command.go:31]: If the buffer gets resized in the host call we should overwrite `buffer` with the returned `buf` so we have the bigger allocated slice prepared for next time.

File: wasm/schema.go [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+import (
+	"fmt"
+
+	conduitv1 "github.com/conduitio/conduit-processor-sdk/proto/conduit/v1"
+	"google.golang.org/protobuf/proto"
+)
+
+func CreateSchema(req *conduitv1.CreateSchemaRequest) (*conduitv1.CreateSchemaResponse, error) {
+	buf, err := proto.Marshal(req)
```
Comment (lovromazgon) [inline wasm/schema.go:27]: I think we could reuse the same global `buffer` 🤔 Or maybe have a global pool of buffers that we reuse. That would maybe be the safer option.

**Review by lovromazgon — APPROVED** [review]
Looking good, let's add the in-memory service in the next PR 👍

**Pull request merged**